File: MergeSort/MergeSort.cs
namespace Struct.MergeSort;

class MergeSort
{
    /*
    Merge function merges two subarrays of arr[]. 
    First subarray is arr[left..mid] 
    Second subarray is arr[mid+1..right]
    Parameters:
       arr: The array containing the subarrays to be merged.
       left: The starting index of the first subarray.
       mid: The ending index of the first subarray.
       right: The ending index of the second subarray.
     */
    static void Merge(int[] arr, int left, int mid, int right)
    {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] leftArr = new int[n1];
        int[] rightArr = new int[n2];

        for (int i = 0; i < n1; ++i)
            leftArr[i] = arr[left + i];

        for (int j = 0; j < n2; ++j)
            rightArr[j] = arr[mid + 1 + j];

        int k = left;
        int i1 = 0;
        int i2 = 0;

        while (i1 < n1 && i2 < n2)
        {
            if (leftArr[i1] <= rightArr[i2])
            {
                arr[k] = leftArr[i1];
                i1++;
            }
            else
            {
                arr[k] = rightArr[i2];
                i2++;
            }

            k++;
        }

        while (i1 < n1)
        {
            arr[k] = leftArr[i1];
            i1++;
            k++;
        }

        while (i2 < n2)
        {
            arr[k] = rightArr[i2];
            i2++;
            k++;
        }
    }
    /*
    Sorts an array using the merge sort algorithm.
    Parameters:
       arr: The array to be sorted.
       left: The starting index of the subarray.
       right: The ending index of the subarray.
    Returns:
       None.
     */

    static void MergeSortAlgorithm(int[] arr, int left, int right)
    {
        if (left < right)
        {
            int mid = left + (right - left) / 2;

            MergeSortAlgorithm(arr, left, mid);
            MergeSortAlgorithm(arr, mid + 1, right);

            Merge(arr, left, mid, right);
        }
    }
}

File: MergeSort/merge_sort.py
"""
Sorts an array using the merge sort algorithm.
Parameters:
- arr (list): The array to be sorted.
Returns:
- list: The sorted array.
This function recursively divides the array into two halves,
sorts each half separately using merge sort, and then merges
the two sorted halves to produce the final sorted array.
"""


def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    left = merge_sort(left)
    right = merge_sort(right)

    return merge(left, right)


"""
Merge two sorted lists into a single sorted list.
Parameters:
    left (List): The first sorted list.
    right (List): The second sorted list.
Returns:
    List: The merged sorted list.
"""


def merge(left, right):
    merged = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1

    while i < len(left):
        merged.append(left[i])
        i += 1

    while j < len(right):
        merged.append(right[j])
        j += 1

    return merged